Mathlib/LinearAlgebra/DirectSum/Finsupp.lean

```diff
@@ -12,20 +12,264 @@ import Mathlib.LinearAlgebra.DirectSum.TensorProduct
 /-!
 # Results on finitely supported functions.
 
-The tensor product of `ι →₀ M` and `κ →₀ N` is linearly equivalent to `(ι × κ) →₀ (M ⊗ N)`.
+* `TensorProduct.finsuppLeft`, the tensor product of `ι →₀ M` and `N`
+  is linearly equivalent to `ι →₀ M ⊗[R] N`
+
+* `TensorProduct.finsuppScalarLeft`, the tensor product of `ι →₀ R` and `N`
+  is linearly equivalent to `ι →₀ N`
+
+* `TensorProduct.finsuppRight`, the tensor product of `M` and `ι →₀ N`
+  is linearly equivalent to `ι →₀ M ⊗[R] N`
+
+* `TensorProduct.finsuppScalarRight`, the tensor product of `M` and `ι →₀ R`
+  is linearly equivalent to `ι →₀ N`
+
+
+* `TensorProduct.finsuppLeft'`, if `M` is an `S`-module,
+  then the tensor product of `ι →₀ M` and `N` is `S`-linearly equivalent
+  to `ι →₀ M ⊗[R] N`
+
+* `finsuppTensorFinsupp`, the tensor product of `ι →₀ M` and `κ →₀ N`
+  is linearly equivalent to `(ι × κ) →₀ (M ⊗ N)`.
+
+## Case of MvPolynomial
+
+These functions apply to `MvPolynomial`, one can define
+```
+noncomputable def MvPolynomial.rTensor' :
+    MvPolynomial σ S ⊗[R] N ≃ₗ[S] (σ →₀ ℕ) →₀ (S ⊗[R] N) :=
+  TensorProduct.finsuppLeft'
+
+noncomputable def MvPolynomial.rTensor :
+    MvPolynomial σ R ⊗[R] N ≃ₗ[R] (σ →₀ ℕ) →₀ N :=
+  TensorProduct.finsuppScalarLeft
+ ```
+
+## Case of `Polynomial`
+
+`Polynomial` is a structure containing a `Finsupp`, so these functions
+can't be applied directly to `Polynomial`.
+
+Some linear equivs need to be added to mathlib for that.
+
+## TODO
+
+* generalize to `MonoidAlgebra`, `AlgHom `
+
+* reprove `TensorProduct.finsuppLeft'` using existing heterobasic version of `TensorProduct.congr`
 -/
 
 
 noncomputable section
 
-open DirectSum Set LinearMap Submodule TensorProduct
+open DirectSum TensorProduct
+
+open Set LinearMap Submodule
+
+section TensorProduct
+
+variable {R : Type*} [CommSemiring R]
+  {M : Type*} [AddCommMonoid M] [Module R M]
+  {N : Type*} [AddCommMonoid N] [Module R N]
+
+namespace TensorProduct
+
+variable {ι : Type*} [DecidableEq ι]
+
+/-- The tensor product of `ι →₀ M` and `N` is linearly equivalent to `ι →₀ M ⊗[R] N` -/
+noncomputable def finsuppLeft :
+    (ι →₀ M) ⊗[R] N ≃ₗ[R] ι →₀ (M ⊗[R] N) :=
+  (congr (finsuppLEquivDirectSum R M ι) (LinearEquiv.refl R N)).trans
+    ((directSumLeft R (fun _ : ι => M) N).trans
+      (finsuppLEquivDirectSum R (M ⊗[R] N) ι).symm)
+
+lemma finsuppLeft_apply_tmul (p : ι →₀ M) (n : N) :
+    finsuppLeft (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m ⊗ₜ[R] n)) := by
+  simp only [finsuppLeft, LinearEquiv.trans_apply, congr_tmul, LinearEquiv.refl_apply]
+  conv_lhs => rw [← Finsupp.sum_single p]
+  rw [LinearEquiv.symm_apply_eq]
+  simp only [map_finsupp_sum]
+  simp only [finsuppLEquivDirectSum_single]
+  rw [← LinearEquiv.eq_symm_apply]
+  simp only [map_finsupp_sum]
+  simp only [directSumLeft_symm_lof_tmul]
+  simp only [Finsupp.sum, sum_tmul]
+
+lemma finsuppLeft_apply_tmul_apply (p : ι →₀ M) (n : N) (i : ι) :
+    finsuppLeft (p ⊗ₜ[R] n) i = p i ⊗ₜ[R] n := by
```

Comment:
Is it a good idea to mark `finsuppLeft_apply_tmul_apply`, `finsuppLeft_symm_apply_single`, `finsuppRight_apply_tmul_apply`, `finsuppRight_symm_apply_single`, `finsuppScalarLeft_apply_tmul_apply`, `finsuppScalarLeft_symm_apply_single`, `finsuppScalarRight_apply_tmul_apply`, `finsuppScalarRight_symm_apply_single` as simp lemmas? Just like the existing `finsuppTensorFinsupp_single`, `finsuppTensorFinsupp_apply` and `finsuppTensorFinsupp_symm_single`.

Reply:
I have no opinion.

Reply:
In #11635 , I marked them with simp